Introduce ThreadSafeWeakHashSet

achristensen07 opened the pull request:
#### 1d67c0b3b7db5d76ebf7743380d9b5f289689902
<pre>
Introduce ThreadSafeWeakHashSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=249005">https://bugs.webkit.org/show_bug.cgi?id=249005</a>

Reviewed by Chris Dumez.

It&apos;s like WeakHashSet but thread safe!  This will allow us to expand our use of
ThreadSafeWeakPtr to classes that used to use WeakHashSet.  I tried hard to use
SFINAE to make WeakHashSet just know if it&apos;s thread safe or not, but cases like
NetworkSession.m_dataTaskSet made that not work.  It needs to instantiate the
template based on only a forward declaration of class NetworkDataTask.  So we
need a different type.  It&apos;s also a fundamentally different structure, with
things like computeSize being immediately useless.  Iteration also needs to
happen differently: we take a thread-safe snapshot of all the members of the
set, then keep them alive during iteration.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/ThreadSafeWeakHashSet.h: Added.
(WTF::copyToVector):
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
* Source/WTF/wtf/WeakHashSet.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.cpp:
(WebCore::FileSystemDirectoryHandle::getFileHandle):
(WebCore::FileSystemDirectoryHandle::getDirectoryHandle):
(WebCore::FileSystemDirectoryHandle::getHandle):
* Source/WebCore/Modules/filesystemaccess/FileSystemHandle.h:
* Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.h:
* Source/WebCore/Modules/webaudio/AudioWorkletProcessor.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::compressedVideoOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriter::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriter::flushCompressedSampleBuffers):
(WebCore::MediaRecorderPrivateWriter::stopRecording):
(WebCore::MediaRecorderPrivateWriter::fetchData):
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::scheduleFailure):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::~WebSWServerToContextConnection):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::acquireSync):
* Source/WebKit/UIProcess/Notifications/WebNotification.h:
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/257817@main">https://commits.webkit.org/257817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90191c5c0d722db4edd15660ec579ac86d9843bc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100122 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9290 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109457 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/169692 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104117 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10173 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92554 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107347 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105891 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90705 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3061 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86667 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/490 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3031 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29047 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9161 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43364 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89547 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5371 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/4871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20020 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -0,0 +1,174 @@
+
+        bool operator==(const const_iterator& other) const
+        {
+            // Not a complete comparison, but only used to compare an iterator against the end.
```
While this may only be used to compare with the `end` iterator right now, this seems like a risky assumption to make in a comparison operator.

achristensen07 replied:
That's why I added an assertion that the other has an empty array, indicating that it is the end.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -0,0 +1,174 @@
+        {
+            // Not a complete comparison, but only used to compare an iterator against the end.
+            ASSERT_UNUSED(other, !other.m_strongReferences.size());
+            return m_position >= m_strongReferences.size();
```
Why can't we `return m_position == other.m_position;` ? We'd probably use `std::numeric_limits<size_t>::max()` for the position of the `end` iterator. I think we'd also need to set m_position to `std::numeric_limits<size_t>::max()` in operator++ in the case where `m_position` is part the end of `m_strongReferences`.

Alternatively, we could not have an iterator at all and require all call sites to use forEach(), it is unclear there is really a need to expose an iterator, especially if this hash set can get modified from another thread. It wouldn't "look" safe even if you wrote your iterator so that it is.

achristensen07 replied:
I like the idea of removing the iterator completely.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -0,0 +1,174 @@
+        const_iterator(Vector<Ref<T>>&& strongReferences)
+            : m_strongReferences(WTFMove(strongReferences)) { }
+
+        const Vector<Ref<T>> m_strongReferences;
```
Iterators are usually very lightweight objects.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -0,0 +1,174 @@
+        return m_set.contains(value.m_controlBlock);
+    }
+
+    bool computesEmpty() const { return begin() == end(); }
```
Given that begin() allocates a vector, this is not an efficient implementation for figuring out if the container is empty.

cdumez reviewed (commented):
I didn't have time to do a full review yet but made some comments.

Also, do we really have a use case of modifying the set from multiple threads? Or is the set always modified on the same thread but it contains TheadSafeWeakPtrs? 

For the first one, you need locking everywhere like you did. But for the second one, you only need to create strong references and put them in a vector before iterating.

cdumez commented on Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.h:
```diff
@@ -84,7 +84,7 @@ class AudioWorkletGlobalScope final : public WorkletGlobalScope {
     const float m_sampleRate;
     MemoryCompactRobinHoodHashMap<String, RefPtr<JSAudioWorkletProcessorConstructor>> m_processorConstructorMap;
     Lock m_processorsLock;
```
We have double-locking now, once outside the HashSet implementation with this `m_processorsLock` and then again inside the new TheadSafeWeakHashSet implementation.

I believe you should drop the `m_processorsLock` lock now that we use a thread safe HashSet implementation.

cdumez reviewed (approved):
LGTM with one fix.

achristensen07 committed:
Introduce ThreadSafeWeakHashSet
https://bugs.webkit.org/show_bug.cgi?id=249005

Reviewed by Chris Dumez.

It's like WeakHashSet but thread safe!  This will allow us to expand our use of
ThreadSafeWeakPtr to classes that used to use WeakHashSet.  I tried hard to use
SFINAE to make WeakHashSet just know if it's thread safe or not, but cases like
NetworkSession.m_dataTaskSet made that not work.  It needs to instantiate the
template based on only a forward declaration of class NetworkDataTask.  So we
need a different type.  It's also a fundamentally different structure, with
things like computeSize being immediately useless.  Iteration also needs to
happen differently: we take a thread-safe snapshot of all the members of the
set, then keep them alive during iteration.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/ThreadSafeWeakHashSet.h: Added.
(WTF::copyToVector):
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
* Source/WTF/wtf/WeakHashSet.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.cpp:
(WebCore::FileSystemDirectoryHandle::getFileHandle):
(WebCore::FileSystemDirectoryHandle::getDirectoryHandle):
(WebCore::FileSystemDirectoryHandle::getHandle):
* Source/WebCore/Modules/filesystemaccess/FileSystemHandle.h:
* Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.h:
* Source/WebCore/Modules/webaudio/AudioWorkletProcessor.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::compressedVideoOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriter::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateWriter::flushCompressedSampleBuffers):
(WebCore::MediaRecorderPrivateWriter::stopRecording):
(WebCore::MediaRecorderPrivateWriter::fetchData):
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::scheduleFailure):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::~WebSWServerToContextConnection):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::acquireSync):
* Source/WebKit/UIProcess/Notifications/WebNotification.h:
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/257817@main
